File: tests/output/jpa/src/main/javagen/restaurant/jpa_sequence_metamodel/api/server/restaurant/CommandeController.java
////
//// ATTENTION CE FICHIER EST GENERE AUTOMATIQUEMENT !
////

package restaurant.jpa_sequence_metamodel.api.server.restaurant;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;

import jakarta.annotation.Generated;
import jakarta.validation.Valid;

import restaurant.jpa_sequence_metamodel.dtos.restaurant.CommandeItem;
import restaurant.jpa_sequence_metamodel.dtos.restaurant.CommandeRead;
import restaurant.jpa_sequence_metamodel.dtos.restaurant.CommandeWrite;
import restaurant.jpa_sequence_metamodel.dtos.restaurant.ReservationRead;
import restaurant.jpa_sequence_metamodel.dtos.restaurant.ReservationWrite;
import restaurant.jpa_sequence_metamodel.entities.restaurant.StatutCommande;
import restaurant.jpa_sequence_metamodel.enums.restaurant.StatutCommandeCode;

@RequestMapping("api/restaurants")
@Generated("TopModel : https://github.com/klee-contrib/topmodel")
public interface CommandeController {

	/**
	 * Crée une nouvelle commande.
	 * @param commande Commande à créer.
	 *
	 * @return Commande créée.
	 */
	@PostMapping(path = "commandes")
	@Operation(description = "Crée une nouvelle commande")
	CommandeRead addCommande(@io.swagger.v3.oas.annotations.parameters.RequestBody(description = "Commande à créer") @RequestBody @Valid CommandeWrite commande);

	/**
	 * Crée une réservation.
	 * @param reservation Réservation à créer.
	 *
	 * @return Réservation créée.
	 */
	@PostMapping(path = "reservations")
	@Operation(description = "Crée une réservation")
	ReservationRead createReservation(@io.swagger.v3.oas.annotations.parameters.RequestBody(description = "Réservation à créer") @RequestBody @Valid ReservationWrite reservation);

	/**
	 * Supprime une commande.
	 * @param comId Identifiant de la commande.
	 */
	@ResponseStatus(HttpStatus.NO_CONTENT)
	@DeleteMapping(path = "commandes/{comId}")
	@Operation(description = "Supprime une commande")
	void deleteCommande(@Parameter(description = "Identifiant de la commande") @PathVariable("comId") Integer comId);

	/**
	 * Exporte les commandes au format CSV.
	 * @param dateDebut Date et heure de la commande.
	 * @param dateFin Date et heure de la commande.
	 *
	 * @return Fichier CSV des commandes.
	 */
	@PreAuthorize("isAuthenticated()")
	@Operation(description = "Exporte les commandes au format CSV")
	@GetMapping(path = "commandes/export", produces = "application/octet-stream")
	byte[] exportCommandes(@Parameter(description = "Date et heure de la commande") @RequestParam(value = "dateDebut", required = true) LocalDateTime dateDebut, @Parameter(description = "Date et heure de la commande") @RequestParam(value = "dateFin", required = true) LocalDateTime dateFin);

	/**
	 * Charge le détail d'une commande.
	 * @param comId Identifiant de la commande.
	 *
	 * @return Détail de la commande.
	 */
	@GetMapping(path = "commandes/{comId}")
	@Operation(description = "Charge le détail d'une commande")
	CommandeRead getCommande(@Parameter(description = "Identifiant de la commande") @PathVariable("comId") Integer comId);

	/**
	 * Liste toutes les commandes.
	 * @param clientId Client ayant passé la commande.
	 * @param statutCommandeCode Statut de la commande.
	 * @param tableId Table associée à la commande.
	 *
	 * @return Liste des commandes.
	 */
	@GetMapping(path = "commandes")
	@Operation(description = "Liste toutes les commandes")
	List<CommandeItem> getCommandes(@Parameter(description = "Client ayant passé la commande") @RequestParam(value = "clientId", required = true) Integer clientId, @Parameter(description = "Statut de la commande") @RequestParam(value = "statutCommandeCode", required = true) StatutCommandeCode statutCommandeCode, @Parameter(description = "Table associée à la commande") @RequestParam(value = "tableId", required = false) Integer tableId);

	/**
	 * Récupère les commandes par date.
	 * @param dateCommande Date et heure de la commande.
	 *
	 * @return Commandes pour la date spécifiée.
	 */
	@GetMapping(path = "commandes/by-date")
	@Operation(description = "Récupère les commandes par date")
	List<CommandeItem> getCommandesByDate(@Parameter(description = "Date et heure de la commande") @RequestParam(value = "dateCommande", required = true) LocalDateTime dateCommande);

	/**
	 * Liste tous les statuts de commande.
	 *
	 * @return Liste des statuts de commande.
	 */
	@GetMapping(path = "statuts-commande")
	@Operation(description = "Liste tous les statuts de commande")
	List<StatutCommande> getStatutCommandes();

	/**
	 * Met à jour partiellement une commande.
	 * @param comId Identifiant de la commande.
	 * @param commande Données partielles de la commande.
	 *
	 * @return Commande mise à jour.
	 */
	@PatchMapping(path = "commandes/{comId}")
	@Operation(description = "Met à jour partiellement une commande")
	CommandeRead patchCommande(@Parameter(description = "Identifiant de la commande") @PathVariable("comId") Integer comId, @io.swagger.v3.oas.annotations.parameters.RequestBody(description = "Données partielles de la commande") @RequestBody @Valid CommandeWrite commande);

	/**
	 * Met à jour une commande.
	 * @param comId Identifiant de la commande.
	 * @param commande Commande à mettre à jour.
	 *
	 * @return Commande mise à jour.
	 */
	@PutMapping(path = "commandes/{comId}")
	@Operation(description = "Met à jour une commande")
	CommandeRead updateCommande(@Parameter(description = "Identifiant de la commande") @PathVariable("comId") Integer comId, @io.swagger.v3.oas.annotations.parameters.RequestBody(description = "Commande à mettre à jour") @RequestBody @Valid CommandeWrite commande);

	/**
	 * Met à jour uniquement le statut d'une commande.
	 * @param comId Identifiant de la commande.
	 * @param statutCommandeCode Statut de la commande.
	 *
	 * @return Commande avec le statut mis à jour.
	 */
	@PatchMapping(path = "commandes/{comId}/statut")
	@Operation(description = "Met à jour uniquement le statut d'une commande")
	CommandeRead updateCommandeStatut(@Parameter(description = "Identifiant de la commande") @PathVariable("comId") Integer comId, @Parameter(description = "Statut de la commande") @RequestParam(value = "statutCommandeCode", required = true) StatutCommandeCode statutCommandeCode);
}

File: tests/output/csharp/src/TopModel.Sample.Api/Controllers/Restaurant/CommandeController.cs
////
//// ATTENTION, CE FICHIER EST PARTIELLEMENT GENERE AUTOMATIQUEMENT !
////

using Microsoft.AspNetCore.Mvc;

namespace TopModel.Sample.Api.Restaurant;

public class CommandeController : Controller
{
    /// <summary>
    /// Crée une nouvelle commande
    /// </summary>
    /// <param name="commande">Commande à créer</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Commande créée</returns>
    [HttpPost("api/restaurants/commandes")]
    public async Task<CommandeRead> AddCommande([FromBody] CommandeWrite commande, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Crée une réservation
    /// </summary>
    /// <param name="reservation">Réservation à créer</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Réservation créée</returns>
    [HttpPost("api/restaurants/reservations")]
    public async Task<ReservationRead> CreateReservation([FromBody] ReservationWrite reservation, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Supprime une commande
    /// </summary>
    /// <param name="comId">Identifiant de la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Task.</returns>
    [HttpDelete("api/restaurants/commandes/{comId:int}")]
    public async Task DeleteCommande(int comId, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Supprime une commande
    /// </summary>
    /// <param name="commandeItem">Commande item à supprimer dans le body</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Task.</returns>
    [HttpDelete("api/restaurants/commandes")]
    public async Task DeleteCommandeWithBody([FromBody] CommandeItem commandeItem, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Exporte les commandes au format CSV
    /// </summary>
    /// <param name="dateDebut">Date et heure de la commande</param>
    /// <param name="dateFin">Date et heure de la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Fichier CSV des commandes</returns>
    [Produces("application/octet-stream")]
    [Authorize]
    [Authorize(Roles = "ADMIN")]
    [HttpGet("api/restaurants/commandes/export")]
    public async Task<byte[]> ExportCommandes([Required] DateTime? dateDebut = null, [Required] DateTime? dateFin = null, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Charge le détail d'une commande
    /// </summary>
    /// <param name="comId">Identifiant de la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Détail de la commande</returns>
    [HttpGet("api/restaurants/commandes/{comId:int}")]
    public async Task<CommandeRead> GetCommande(int comId, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Liste toutes les commandes
    /// </summary>
    /// <param name="clientId">Client ayant passé la commande</param>
    /// <param name="statutCommandeCode">Statut de la commande</param>
    /// <param name="tableId">Table associée à la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Liste des commandes</returns>
    [HttpGet("api/restaurants/commandes")]
    public async Task<ICollection<CommandeItem>> GetCommandes([Required] int? clientId = null, StatutCommande.Codes statutCommandeCode = StatutCommande.Codes.EN_ATT, int? tableId = null, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Récupère les commandes par date
    /// </summary>
    /// <param name="dateCommande">Date et heure de la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Commandes pour la date spécifiée</returns>
    [HttpGet("api/restaurants/commandes/by-date")]
    public async Task<ICollection<CommandeItem>> GetCommandesByDate([Required] DateTime? dateCommande = null, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Liste tous les statuts de commande
    /// </summary>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Liste des statuts de commande</returns>
    [HttpGet("api/restaurants/statuts-commande")]
    public async Task<ICollection<StatutCommande>> GetStatutCommandes(CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Met à jour partiellement une commande
    /// </summary>
    /// <param name="comId">Identifiant de la commande</param>
    /// <param name="commande">Données partielles de la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Commande mise à jour</returns>
    [HttpPatch("api/restaurants/commandes/{comId:int}")]
    public async Task<CommandeRead> PatchCommande(int comId, [FromBody] CommandeWrite commande, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Met à jour une commande
    /// </summary>
    /// <param name="comId">Identifiant de la commande</param>
    /// <param name="commande">Commande à mettre à jour</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Commande mise à jour</returns>
    [HttpPut("api/restaurants/commandes/{comId:int}")]
    public async Task<CommandeRead> UpdateCommande(int comId, [FromBody] CommandeWrite commande, CancellationToken ct = default)
    {

    }

    /// <summary>
    /// Met à jour uniquement le statut d'une commande
    /// </summary>
    /// <param name="comId">Identifiant de la commande</param>
    /// <param name="statutCommandeCode">Statut de la commande</param>
    /// <param name="ct">CancellationToken (HttpContext.RequestAborted).</param>
    /// <returns>Commande avec le statut mis à jour</returns>
    [HttpPatch("api/restaurants/commandes/{comId:int}/statut")]
    public async Task<CommandeRead> UpdateCommandeStatut(int comId, StatutCommande.Codes statutCommandeCode = StatutCommande.Codes.EN_ATT, CancellationToken ct = default)
    {

    }

}